File: frequencyTopic2020/classical/maxProfit.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int maxProfit(vector<int>& prices) {
    if (prices.empty() || prices.size() == 1) {
      return 0;
    }

    int res = 0;
    int low = prices[0];
    for (int i = 1; i < prices.size();i++) {
      low = min(low, prices[i]);
      res = max(prices[i] - low, res);
    }
    return res;
  }
};

File: rush/dp/maxProfit.go
package main

func maxProfit(prices []int) int {

    profit := 0
    lowest := math.MaxInt64
    for i := 0; i < len(prices); i++ {
        lowest = min(lowest, prices[i])
        profit = max(profit, prices[i] - lowest)
    }

    return profit
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}
